authorization/v2/uniformResourceIdentifier.go

```diff
@@ -157,3 +158,57 @@ func marshallURIMap(uriMap map[string]string) string {
 	// Remove the extra '&' character introduced when marshaling the uriMap
 	return marshalledMap[:len(marshalledMap)-1]
 }
+
+// isURIMatch checks that the target URI is a subset of the source URI (the URI which the user can access)
+func (uri UniformResourceIdentifier) isSubsetOf(target UniformResourceIdentifier) bool {
+	// Ensure the target path is a subset of the source path
+	if len(uri.path) > len(target.path) {
+		return false
+	}
+
+	// Compare URI path
+	for i, sourceCharacter := range uri.path {
+		if uint8(sourceCharacter) != target.path[i] {
+			return false
+		}
+	}
+
+	// Validate URI arguments
+	for key, sourceValue := range uri.arguments {
+		if targetValue, ok := target.arguments[key]; ok {
+			match, err := regexp.Match(sourceValue, []byte(targetValue))
+			if !match || err != nil {
+				// Fail-soft, if the regex is invalid or the regex pattern match fails, the URIs don't match
+				return false
+			}
+		} else {
+			// In the case the target URI doesn't contain an argument in the source
+			// the target is no longer a subset of the source.
+			// e.g. Source = hs:hs_auth?test=1 and Target = hs:hs_auth
+			// this means, hs:hs_auth is not a subset of the source since it is limited
+			// by the argument "test=1"
+			return false
+		}
+	}
+
+	// Validate URI metadata
+	for key, sourceValue := range uri.metadata {
```

Comment:
This is not how metadata on URIs would work.
The metadata determines under which conditions a specific URI is valid not how the URIs it's being compared to should look like. E.g.: `hs:hs_auth:api:v2` would be a subset of `hs:hs_auth:api:v2#until=45546645` as long as the current Unix timestamp is less than 45546645. The `#until=45546645` determines if `hs:hs_auth:api:v2#until=45546645` is a valid URI, not if `hs:hs_auth:api:v2` is a subset of `hs:hs_auth:api:v2#until=45546645`.

Instead of this loop we'd need something like:
```
// checking if uri is a valid URI
for key, value := range uri.metadata {
  if metadataHandlers[key](value) == false {
    return false
  }
}

// checking if target is a valid URI
for key, value := range target.metadata {
  if metadataHandlers[key](value) == false {
    return false
  }
}
```
where `metadataHandlers` would be a `map[string]func(string) bool`

This PR is already rather big, so I think we should implement metadata validation as a separate PR. Could you remove the metadata logic for now and just add a TODO in its place?